From f5c8334068d95ac0457d32693a5ede615a0bebf5 Mon Sep 17 00:00:00 2001
From: Loren Huang <b02279@freescale.com>
Date: Wed, 26 Nov 2014 11:59:38 +0800
Subject: [PATCH] MGS-304 [#1461]GPU driver will cause kernel panic when
 allocate memory failed

-The issue is triggered by suspend/resume test
when doing bonnie++ which will consume lots of
memory.

-The root cause is vivante didn't report the allocation
failure to uplevel correctly which cause the improper
free.

-Correct the free logic to fix this issue.

Date: Nov 26, 2014
Signed-off-by: Loren Huang <b02279@freescale.com>
Acked-by: Jason Liu
Tested-by: Peter Chen
(cherry picked from commit 2378f1c0b48f1c632a96c1e6c1107e2773f34170)
---
 drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
index f5dd13c..2dbd73d 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
@@ -1787,6 +1787,8 @@ OnError:
         /* Free LINUX_MDL. */
         gcmkVERIFY_OK(_DestroyMdl(mdl));
     }
+    *Physical = gcvNULL;
+    *Bytes = 0;
 
     if (locked)
     {
@@ -3720,6 +3722,7 @@ OnError:
         /* Free the memory. */
         _DestroyMdl(mdl);
     }
+    *Physical = gcvNULL;
 
     /* Return the status. */
     gcmkFOOTER_ARG("Os=0x%X Flag=%x Bytes=%lu", Os, Flag, Bytes);
-- 
1.8.0

